### PR TITLE
Add `StencilMatrix.dot` backend propagation

### PR DESCRIPTION
**Pull request by davschneller**
This PR lets the backend for `StencilMatrix.dot` propagate when multiplying, adding or transposing that said `StencilMatrix`.

* Propagates the `backend` parameter to all stencil matrices which are generated by `__mul__`, `__add__`, `copy`, `transpose`, etc. (in the case of `__add__` or `__sub__`, a warning is thrown if the backend of the left-hand side and right-hand side matrices differ; then the backend of the left-hand side is chosen by default)
* Makes it possible to use `PSYDAC_BACKEND_PYTHON` as backend
* `set_backend` can now be called again with another backend, the original args are stored
* Includes tests

### CODE REVIEW

File: psydac/linalg/stencil.py [hunk 1]
```diff
@@ -763,7 +767,7 @@ def __add__(self, m):
         assert m._domain   is self._domain
         assert m._codomain is self._codomain
         assert m._pads     == self._pads
-        w = StencilMatrix(self._domain, self._codomain, self._pads)
+        w = StencilMatrix(self._domain, self._codomain, self._pads, self._backend)
```
Comment (yguclu) [inline psydac/linalg/stencil.py:770]: Can you please check if the two matrices use the same backend, and if they don't raise a clear warning like the following?

```python
import warnings

if m._backend != self._backend:
    msg = 'Adding two matrices with different backends is ambiguous - defaulting to backend of first addend'
    warnings.warn(msg, category=RuntimeWarning)
```

File: psydac/linalg/stencil.py [hunk 2]
```diff
@@ -776,7 +780,7 @@ def __sub__(self, m):
...
```
Comment (yguclu) [inline psydac/linalg/stencil.py:783]: Same comment as above

**Review by yguclu — COMMENTED** [review]
Good job @davschneller!

I would like a runtime warning to be raised if we have an ambiguity in the result's backend.

**Review by yguclu — COMMENTED** [review]
Thanks for the changes. Please update the PR's description...

**Comment by davschneller** [comment]
Yes, it is now updated

**Review by yguclu — APPROVED** [review]
Waiting for [github actions problem](https://www.githubstatus.com/incidents/m16jzl31gnqt) to be fixed

**Pull request merged**